File: moling-server/app/dao/template_dao.py
"""墨灵 (Moling) — Template DAO."""

from __future__ import annotations

from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from app.dao.base_dao import BaseDAO
from app.models.template import Template


class TemplateDAO(BaseDAO[Template]):
    """Data access object for Template model."""

    def __init__(self):
        super().__init__(Template)

    async def get_by_genre(
        self,
        db: AsyncSession,
        genre: str,
        *,
        skip: int = 0,
        limit: int = 20,
    ) -> list[Template]:
        """Get templates by genre."""
        stmt = (
            select(Template)
            .where(Template.genre == genre)
            .order_by(Template.created_at.desc())
            .offset(skip)
            .limit(limit)
        )
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def count_by_genre(
        self,
        db: AsyncSession,
        genre: str,
    ) -> int:
        """Count templates by genre."""
        stmt = select(func.count()).select_from(Template).where(Template.genre == genre)
        result = await db.execute(stmt)
        return result.scalar_one()


# Singleton instance
template_dao = TemplateDAO()

File: moling-server-rs/crates/moling-db/src/dao/template_dao.rs
//! Template DAO — template CRUD with genre-based queries.

use moling_core::error::{AppError, AppResult};
use sea_orm::{ActiveModelTrait, ColumnTrait, DatabaseConnection, EntityTrait, PaginatorTrait, QueryFilter, QueryOrder, QuerySelect};

use crate::entities::template::{self, ActiveModel, Entity as Template, Model as TemplateModel};

/// Template data access object.
#[derive(Clone, Default)]
pub struct TemplateDao;

impl TemplateDao {
    /// Find a template by ID.
    pub async fn find_by_id(
        &self,
        db: &DatabaseConnection,
        id: &str,
    ) -> AppResult<Option<TemplateModel>> {
        Template::find_by_id(id.to_owned())
            .one(db)
            .await
            .map_err(|e| {
                tracing::error!(%id, "Template: database error finding by id: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// List templates by genre with pagination.
    pub async fn list_by_genre(
        &self,
        db: &DatabaseConnection,
        genre: &str,
        skip: u64,
        limit: u64,
    ) -> AppResult<Vec<TemplateModel>> {
        Template::find()
            .filter(template::Column::Genre.eq(genre))
            .order_by_desc(template::Column::CreatedAt)
            .offset(skip)
            .limit(limit)
            .all(db)
            .await
            .map_err(|e| {
                tracing::error!(%genre, "Template: database error listing by genre: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Count templates by genre.
    pub async fn count_by_genre(
        &self,
        db: &DatabaseConnection,
        genre: &str,
    ) -> AppResult<u64> {
        Template::find()
            .filter(template::Column::Genre.eq(genre))
            .count(db)
            .await
            .map_err(|e| {
                tracing::error!(%genre, "Template: database error counting by genre: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Create a new template.
    pub async fn create(
        &self,
        db: &DatabaseConnection,
        model: ActiveModel,
    ) -> AppResult<TemplateModel> {
        model.insert(db).await.map_err(|e| {
            tracing::error!("Template: database error creating: {e}");
            AppError::internal("Database insert failed")
        })
    }

    /// Update an existing template.
    pub async fn update(
        &self,
        db: &DatabaseConnection,
        model: ActiveModel,
    ) -> AppResult<TemplateModel> {
        model.update(db).await.map_err(|e| {
            tracing::error!("Template: database error updating: {e}");
            AppError::internal("Database update failed")
        })
    }

    /// Hard-delete a template.
    pub async fn delete(&self, db: &DatabaseConnection, id: &str) -> AppResult<u64> {
        let result = Template::delete_by_id(id.to_owned())
            .exec(db)
            .await
            .map_err(|e| {
                tracing::error!(%id, "Template: database error deleting: {e}");
                AppError::internal("Database delete failed")
            })?;
        Ok(result.rows_affected)
    }
}
